core/src/HIP/Kokkos_HIP_Space.cpp

```diff
@@ -43,6 +43,25 @@ namespace {
 
 static std::atomic<bool> is_first_hip_managed_allocation(true);
 
+auto get_failure_mode(hipError_t error_code) {
+  using FailureMode =
+      Kokkos::Experimental::RawMemoryAllocationFailure::FailureMode;
+  switch (error_code) {
+    case hipErrorMemoryAllocation: return FailureMode::OutOfMemoryError;
+    case hipErrorInvalidValue: return FailureMode::InvalidAllocationSize;
+    default: return FailureMode::Unknown;
+  }
+}
+
+void throw_hip_allocation_failure(size_t alloc_size, hipError_t error_code,
+                                  std::string msg) {
+  msg += "returned error code \"";
+  msg += hipGetErrorName(error_code);
+  msg += "\"";
+  Kokkos::Impl::throw_bad_alloc(alloc_size, std::align_val_t{1},
+                                get_failure_mode(error_code), std::move(msg));
```

Comment:
At least add a FIXME for the alignment? Ultimately, we at least assume an alignment of 64 byes throughout the code base anyway.